metric_learn/constraints.py

```diff
@@ -33,6 +35,110 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
+    """
+    Generates triplets from labeled data.
+
+    For every point (X_a) the triplets (X_a, X_b, X_c) are constructed from all
+    the combinations of taking one of its `k_genuine`-nearest neighbors of the
+    same class (X_b) and taking one of its `k_impostor`-nearest neighbors of
+    other classes (X_c).
+
+    In the case a class doesn't have enough points in the same class (other
+    classes) to yield `k_genuine` (`k_impostor`) neighbors a warning will be
+    raised and the maximum value of genuine (impostor) neighbors will be used
+    for that class.
+
+    Parameters
+    ----------
+      X : (n x d) matrix
+        Input data, where each row corresponds to a single instance.
+      k_genuine : int
+        Number of neighbors of the same class to be taken into account.
+      k_impostor : int
+        Number of neighbors of different classes to be taken into account.
+
+    Returns
+    -------
+    triplets : array-like, shape=(n_constraints, 3)
+      2D array of triplets of indicators.
+    """
+
+    labels, labels_count = np.unique(self.partial_labels, return_counts=True)
+    n_labels = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+
+    # Handle the case where there are too few elements to yield k_genuine or
+    # k_impostor neighbors for every class.
+
+    k_genuine_vec = np.ones(n_labels, dtype=np.intp)*k_genuine
```

Comment:
`np.full_like(labels, k_genuine)`

Same for the next line

Reply:
maybe `np.full(n_labels, k_genuine)` is better?

Reply:
I prefer `full_like` as it's more explicit that we're making it the same shape as the `labels` array. Ideally we wouldn't need to store `n_labels` anywhere.

You can also pass a `dtype=np.intp` to `full_like`, though I suspect the labels array is already of integer type.

Reply:
Alright! You are correct `n_labels` isn't needed anymore. The `self.partial_labels` is indeed initialized as `int` in the `Constraints.__init__()` 

Should we handle unknown labels? (`label = -1`) I am thinking this could be easily handled by restricting `labels, labels_count` to the mask `labels!=-1`. This way the unknown labeled will be used for the impostor neighbors.

Reply:
I think we should just ignore the unlabeled points (all points with negative label) as done in the `pairs` method. A point with unknown label may have the same label as the target point, so I don't think we should allow them as candidates for impostors.

Reply:
You are right! I will do that then.

Reply:
Maybe you can add one unlabeled point to your manually defined dataset (one that is closer to some other point) so that the test verified that it is ignored

Reply:
Done!